File: android/Gutenberg/src/test/java/org/wordpress/gutenberg/HttpServerAuthenticationTests.kt
package org.wordpress.gutenberg

import org.junit.After
import org.junit.Assert.assertEquals
import org.junit.Before
import org.junit.Test
import java.net.HttpURLConnection
import java.net.Socket
import java.net.URL

class HttpServerAuthenticationTests {

    private lateinit var server: HttpServer

    @Before
    fun setUp() {
        server = HttpServer(
            name = "auth-test",
            externallyAccessible = false,
            requiresAuthentication = true,
            handler = { HttpResponse(body = "OK\n".toByteArray()) }
        )
        server.start()
    }

    @After
    fun tearDown() {
        server.stop()
    }

    @Test
    fun `request without token returns 407 with Content-Type and Proxy-Authenticate`() {
        val conn = URL("http://127.0.0.1:${server.port}/test").openConnection() as HttpURLConnection
        try {
            assertEquals(407, conn.responseCode)
            assertEquals("text/plain", conn.getHeaderField("Content-Type"))
            assertEquals("Bearer", conn.getHeaderField("Proxy-Authenticate"))
        } finally {
            conn.disconnect()
        }
    }

    @Test
    fun `request with wrong token returns 407`() {
        val response = send(server, listOf("Proxy-Authorization" to "Bearer wrong-token"))
        assertEquals(407, response.status)
        assertEquals("Bearer", response.headers["proxy-authenticate"])
    }

    @Test
    fun `request with valid token returns 200`() {
        assertEquals(200, send(server, listOf("Proxy-Authorization" to "Bearer ${server.token}")).status)
    }

    @Test
    fun `request with lowercase 'bearer' scheme returns 200`() {
        assertEquals(200, send(server, listOf("Proxy-Authorization" to "bearer ${server.token}")).status)
    }

    @Test
    fun `request with uppercase 'BEARER' scheme returns 200`() {
        assertEquals(200, send(server, listOf("Proxy-Authorization" to "BEARER ${server.token}")).status)
    }

    // Relay-Authorization (fetch()-compatible alternative)

    @Test
    fun `Relay-Authorization with valid token returns 200`() {
        val conn = URL("http://127.0.0.1:${server.port}/test").openConnection() as HttpURLConnection
        conn.setRequestProperty("Relay-Authorization", "Bearer ${server.token}")
        try {
            assertEquals(200, conn.responseCode)
        } finally {
            conn.disconnect()
        }
    }

    @Test
    fun `Relay-Authorization with wrong token returns 407`() {
        val conn = URL("http://127.0.0.1:${server.port}/test").openConnection() as HttpURLConnection
        conn.setRequestProperty("Relay-Authorization", "Bearer wrong-token")
        try {
            assertEquals(407, conn.responseCode)
        } finally {
            conn.disconnect()
        }
    }

    @Test
    fun `Relay-Authorization with lowercase 'bearer' scheme returns 200`() {
        val conn = URL("http://127.0.0.1:${server.port}/test").openConnection() as HttpURLConnection
        conn.setRequestProperty("Relay-Authorization", "bearer ${server.token}")
        try {
            assertEquals(200, conn.responseCode)
        } finally {
            conn.disconnect()
        }
    }

    @Test
    fun `Authorization header passes through to handler alongside Relay-Authorization`() {
        server.stop()

        var receivedAuth: String? = null
        val authServer = HttpServer(
            name = "auth-test-relay-passthrough",
            externallyAccessible = false,
            requiresAuthentication = true,
            handler = { request ->
                receivedAuth = request.header("Authorization")
                HttpResponse(body = "OK\n".toByteArray())
            }
        )
        authServer.start()
        try {
            val conn = URL("http://127.0.0.1:${authServer.port}/test").openConnection() as HttpURLConnection
            conn.setRequestProperty("Relay-Authorization", "Bearer ${authServer.token}")
            conn.setRequestProperty("Authorization", "Basic dXNlcjpwYXNz")
            try {
                assertEquals(200, conn.responseCode)
                assertEquals("Basic dXNlcjpwYXNz", receivedAuth)
            } finally {
                conn.disconnect()
            }
        } finally {
            authServer.stop()
        }
    }

    @Test
    fun `Proxy-Authorization takes precedence over Relay-Authorization`() {
        val response = send(
            server,
            listOf(
                "Proxy-Authorization" to "Bearer ${server.token}",
                "Relay-Authorization" to "Bearer wrong"
            )
        )
        assertEquals(200, response.status)
    }

    // Authorization Passthrough

    @Test
    fun `Authorization header passes through to handler alongside Proxy-Authorization`() {
        var receivedAuth: String? = null

        server.stop()
        val authServer = HttpServer(
            name = "auth-test-passthrough",
            externallyAccessible = false,
            requiresAuthentication = true,
            handler = { request ->
                receivedAuth = request.header("Authorization")
                HttpResponse(body = "OK\n".toByteArray())
            }
        )
        authServer.start()
        try {
            val response = send(
                authServer,
                listOf(
                    "Proxy-Authorization" to "Bearer ${authServer.token}",
                    "Authorization" to "Basic dXNlcjpwYXNz"
                )
            )
            assertEquals(200, response.status)
            assertEquals("Basic dXNlcjpwYXNz", receivedAuth)
        } finally {
            authServer.stop()
        }
    }

    // CORS Preflight (OPTIONS) Auth Exemption

    @Test
    fun `OPTIONS without token returns 200 (CORS preflight exempt from auth)`() {
        assertEquals(200, send(server, emptyList(), method = "OPTIONS").status)
    }

    @Test
    fun `GET without token still returns 407 (only OPTIONS is exempt)`() {
        val conn = URL("http://127.0.0.1:${server.port}/test").openConnection() as HttpURLConnection
        try {
            assertEquals(407, conn.responseCode)
        } finally {
            conn.disconnect()
        }
    }

    // Content-Length Requirement

    @Test
    fun `POST without Content-Length returns 411`() {
        val response = send(
            server,
            listOf("Proxy-Authorization" to "Bearer ${server.token}"),
            method = "POST"
        )
        assertEquals(411, response.status)
    }

    @Test
    fun `GET without Content-Length returns 200`() {
        assertEquals(200, send(server, listOf("Proxy-Authorization" to "Bearer ${server.token}")).status)
    }

    @Test
    fun `POST with Content-Length returns 200`() {
        val response = send(
            server,
            listOf("Proxy-Authorization" to "Bearer ${server.token}"),
            method = "POST",
            body = "hello".toByteArray()
        )
        assertEquals(200, response.status)
    }

    // Oversized Payloads (auth precedes drain)

    @Test
    fun `oversized request without token returns 407, not 413`() {
        val smallServer = oversizedTestServer()
        try {
            // Auth is checked on headers alone, before the oversized body is
            // drained or the handler runs — so the request is rejected with
            // 407, not answered with the library's 413. An unauthenticated
            // client must not be able to make the server read (and discard)
            // an arbitrarily large body. The body is declared but never sent,
            // so a server that read it before checking auth would never answer.
            val response = send(
                smallServer,
                emptyList(),
                method = "POST",
                contentLength = OVERSIZED_BODY_SIZE
            )
            assertEquals(407, response.status)
        } finally {
            smallServer.stop()
        }
    }

    @Test
    fun `oversized request with valid token is answered 413 by the library, bypassing the handler`() {
        val smallServer = oversizedTestServer()
        try {
            // The library answers a recoverable parse error itself; the handler
            // (which would return 200 "OK") is never invoked for a rejected
            // request.
            val response = send(
                smallServer,
                listOf("Proxy-Authorization" to "Bearer ${smallServer.token}"),
                method = "POST",
                body = ByteArray(OVERSIZED_BODY_SIZE)
            )
            assertEquals(413, response.status)
        } finally {
            smallServer.stop()
        }
    }

    /**
     * Sends a request to [target] over a raw socket. `HttpURLConnection` can't
     * send `Proxy-Authorization`: since JDK-8384708 it strips that header from
     * non-proxied connections.
     *
     * Sends `Content-Length` only when [contentLength] is set, which defaults to
     * [body]'s size. Headers are flushed before the body so the server can act on
     * them before the body arrives, as it would with a real client.
     */
    private fun send(
        target: HttpServer,
        headers: List<Pair<String, String>>,
        method: String = "GET",
        body: ByteArray? = null,
        contentLength: Int? = body?.size
    ): RawResponse {
        Socket("127.0.0.1", target.port).use { sock ->
            sock.soTimeout = SOCKET_TIMEOUT_MS
            val head = buildString {
                append("$method /test HTTP/1.1\r\nHost: 127.0.0.1\r\n")
                headers.forEach { (name, value) -> append("$name: $value\r\n") }
                if (contentLength != null) append("Content-Length: $contentLength\r\n")
                append("\r\n")
            }
            val output = sock.getOutputStream()
            output.write(head.toByteArray())
            output.flush()
            body?.let {
                output.write(it)
                output.flush()
            }

            val reader = sock.getInputStream().bufferedReader()
            val status = reader.readLine().split(" ")[1].toInt()
            val responseHeaders = generateSequence { reader.readLine() }
                .takeWhile { it.isNotEmpty() }
                .associate { line ->
                    val (name, value) = line.split(":", limit = 2)
                    name.trim().lowercase() to value.trim()
                }
            return RawResponse(status, responseHeaders)
        }
    }

    /** A response's status code and headers, keyed by lowercase name. */
    private class RawResponse(val status: Int, val headers: Map<String, String>)

    /** A server whose 1 KB body limit lets a 2 KB POST exercise the drain path. Its
     *  handler only ever answers valid requests — a rejected (oversized) request is
     *  answered by the library, not here. */
    private fun oversizedTestServer(): HttpServer {
        val smallServer = HttpServer(
            name = "auth-drain-test",
            externallyAccessible = false,
            requiresAuthentication = true,
            maxBodySize = 1024L,
            handler = { HttpResponse(body = "OK\n".toByteArray()) }
        )
        smallServer.start()
        return smallServer
    }

    // Auth Disabled

    @Test
    fun `authentication disabled passes through without token`() {
        server.stop()

        val noAuthServer = HttpServer(
            name = "auth-test-no-auth",
            externallyAccessible = false,
            requiresAuthentication = false,
            handler = { HttpResponse(body = "OK\n".toByteArray()) }
        )
        noAuthServer.start()
        try {
            val conn = URL("http://127.0.0.1:${noAuthServer.port}/test").openConnection() as HttpURLConnection
            try {
                assertEquals(200, conn.responseCode)
            } finally {
                conn.disconnect()
            }
        } finally {
            noAuthServer.stop()
        }
    }

    companion object {
        /** Twice the oversized test server's 1 KB `maxBodySize`. */
        private const val OVERSIZED_BODY_SIZE = 2048

        /** Fails a test that gets no response rather than hanging it. */
        private const val SOCKET_TIMEOUT_MS = 5000
    }
}

File: ios/Tests/GutenbergKitHTTPTests/HTTPServerAuthenticationTests.swift
#if canImport(Network)

import Foundation
import Network
import Testing
@testable import GutenbergKitHTTP

@Suite("HTTPServer Authentication")
struct HTTPServerAuthenticationTests {

    @Test("request without token returns 407 with Content-Type and Proxy-Authenticate")
    func noTokenReturns407WithHeaders() async throws {
        let server = try await HTTPServer.start(
            name: "auth-test",
            requiresAuthentication: true
        ) { _ in
            HTTPResponse(status: 200, body: Data("OK\n".utf8))
        }
        defer { server.stop() }

        let url = URL(string: "http://127.0.0.1:\(server.port)/test")!
        let (_, response) = try await URLSession.shared.data(for: URLRequest(url: url))
        let http = try #require(response as? HTTPURLResponse)

        #expect(http.statusCode == 407)
        #expect(http.value(forHTTPHeaderField: "Content-Type") == "text/plain")
        #expect(http.value(forHTTPHeaderField: "Proxy-Authenticate") == "Bearer")
    }

    @Test("request with wrong token returns 407")
    func wrongTokenReturns407() async throws {
        let server = try await HTTPServer.start(
            name: "auth-test",
            requiresAuthentication: true
        ) { _ in
            HTTPResponse(status: 200, body: Data("OK\n".utf8))
        }
        defer { server.stop() }

        let url = URL(string: "http://127.0.0.1:\(server.port)/test")!
        var request = URLRequest(url: url)
        request.setValue("Bearer wrong-token", forHTTPHeaderField: "Proxy-Authorization")

        let (_, response) = try await URLSession.shared.data(for: request)
        let http = try #require(response as? HTTPURLResponse)

        #expect(http.statusCode == 407)
        #expect(http.value(forHTTPHeaderField: "Proxy-Authenticate") == "Bearer")
    }

    @Test("request with valid token returns 200")
    func validTokenReturns200() async throws {
        let server = try await HTTPServer.start(
            name: "auth-test",
            requiresAuthentication: true
        ) { _ in
            HTTPResponse(status: 200, body: Data("OK\n".utf8))
        }
        defer { server.stop() }

        let url = URL(string: "http://127.0.0.1:\(server.port)/test")!
        var request = URLRequest(url: url)
        request.setValue("Bearer \(server.token)", forHTTPHeaderField: "Proxy-Authorization")

        let (_, response) = try await URLSession.shared.data(for: request)
        let http = try #require(response as? HTTPURLResponse)

        #expect(http.statusCode == 200)
    }

    @Test("request with lowercase 'bearer' scheme returns 200")
    func lowercaseBearerSchemeReturns200() async throws {
        let server = try await HTTPServer.start(
            name: "auth-test",
            requiresAuthentication: true
        ) { _ in
            HTTPResponse(status: 200, body: Data("OK\n".utf8))
        }
        defer { server.stop() }

        let url = URL(string: "http://127.0.0.1:\(server.port)/test")!
        var request = URLRequest(url: url)
        request.setValue("bearer \(server.token)", forHTTPHeaderField: "Proxy-Authorization")

        let (_, response) = try await URLSession.shared.data(for: request)
        let http = try #require(response as? HTTPURLResponse)

        #expect(http.statusCode == 200)
    }

    @Test("request with uppercase 'BEARER' scheme returns 200")
    func uppercaseBearerSchemeReturns200() async throws {
        let server = try await HTTPServer.start(
            name: "auth-test",
            requiresAuthentication: true
        ) { _ in
            HTTPResponse(status: 200, body: Data("OK\n".utf8))
        }
        defer { server.stop() }

        let url = URL(string: "http://127.0.0.1:\(server.port)/test")!
        var request = URLRequest(url: url)
        request.setValue("BEARER \(server.token)", forHTTPHeaderField: "Proxy-Authorization")

        let (_, response) = try await URLSession.shared.data(for: request)
        let http = try #require(response as? HTTPURLResponse)

        #expect(http.statusCode == 200)
    }

    // MARK: - Relay-Authorization (fetch()-compatible alternative)

    @Test("Relay-Authorization with valid token returns 200")
    func relayAuthValidTokenReturns200() async throws {
        let server = try await HTTPServer.start(
            name: "auth-test",
            requiresAuthentication: true
        ) { _ in
            HTTPResponse(status: 200, body: Data("OK\n".utf8))
        }
        defer { server.stop() }

        let url = URL(string: "http://127.0.0.1:\(server.port)/test")!
        var request = URLRequest(url: url)
        request.setValue("Bearer \(server.token)", forHTTPHeaderField: "Relay-Authorization")

        let (_, response) = try await URLSession.shared.data(for: request)
        let http = try #require(response as? HTTPURLResponse)

        #expect(http.statusCode == 200)
    }

    @Test("Relay-Authorization with wrong token returns 407")
    func relayAuthWrongTokenReturns407() async throws {
        let server = try await HTTPServer.start(
            name: "auth-test",
            requiresAuthentication: true
        ) { _ in
            HTTPResponse(status: 200, body: Data("OK\n".utf8))
        }
        defer { server.stop() }

        let url = URL(string: "http://127.0.0.1:\(server.port)/test")!
        var request = URLRequest(url: url)
        request.setValue("Bearer wrong-token", forHTTPHeaderField: "Relay-Authorization")

        let (_, response) = try await URLSession.shared.data(for: request)
        let http = try #require(response as? HTTPURLResponse)

        #expect(http.statusCode == 407)
    }

    @Test("Relay-Authorization with lowercase 'bearer' scheme returns 200")
    func relayAuthLowercaseBearerReturns200() async throws {
        let server = try await HTTPServer.start(
            name: "auth-test",
            requiresAuthentication: true
        ) { _ in
            HTTPResponse(status: 200, body: Data("OK\n".utf8))
        }
        defer { server.stop() }

        let url = URL(string: "http://127.0.0.1:\(server.port)/test")!
        var request = URLRequest(url: url)
        request.setValue("bearer \(server.token)", forHTTPHeaderField: "Relay-Authorization")

        let (_, response) = try await URLSession.shared.data(for: request)
        let http = try #require(response as? HTTPURLResponse)

        #expect(http.statusCode == 200)
    }

    @Test("Authorization header passes through to handler alongside Relay-Authorization")
    func authorizationPassesThroughWithRelayAuth() async throws {
        let server = try await HTTPServer.start(
            name: "auth-test",
            requiresAuthentication: true
        ) { req in
            let auth = req.parsed.header("Authorization") ?? ""
            let relayAuth = req.parsed.header("Relay-Authorization") ?? "absent"
            return HTTPResponse(
                status: 200,
                headers: [("X-Received-Auth", auth), ("X-Received-Relay", relayAuth)],
                body: Data("OK\n".utf8)
            )
        }
        defer { server.stop() }

        let raw = "GET /test HTTP/1.1\r\nHost: 127.0.0.1\r\nRelay-Authorization: Bearer \(server.token)\r\nAuthorization: Basic dXNlcjpwYXNz\r\n\r\n"
        let response = try await sendRaw(raw, toPort: server.port)

        #expect(response.contains("HTTP/1.1 200"))
        #expect(response.contains("X-Received-Auth: Basic dXNlcjpwYXNz"))
    }

    @Test("Proxy-Authorization takes precedence over Relay-Authorization")
    func proxyAuthTakesPrecedenceOverRelayAuth() async throws {
        let server = try await HTTPServer.start(
            name: "auth-test",
            requiresAuthentication: true
        ) { _ in
            HTTPResponse(status: 200, body: Data("OK\n".utf8))
        }
        defer { server.stop() }

        // Proxy-Authorization has the correct token, Relay-Authorization has a wrong one.
        // Server should accept because Proxy-Authorization is checked first.
        let raw = "GET /test HTTP/1.1\r\nHost: 127.0.0.1\r\nProxy-Authorization: Bearer \(server.token)\r\nRelay-Authorization: Bearer wrong\r\n\r\n"
        let response = try await sendRaw(raw, toPort: server.port)
        #expect(response.hasPrefix("HTTP/1.1 200"))
    }

    // MARK: - Authorization Passthrough

    @Test("Authorization header passes through to handler alongside Proxy-Authorization")
    func authorizationPassesThroughToHandler() async throws {
        let server = try await HTTPServer.start(
            name: "auth-test",
            requiresAuthentication: true
        ) { req in
            // Echo the Authorization header back in X-Received-Auth so the
            // test can verify it arrived at the handler untouched.
            let auth = req.parsed.header("Authorization") ?? ""
            return HTTPResponse(
                status: 200,
                headers: [("X-Received-Auth", auth)],
                body: Data("OK\n".utf8)
            )
        }
        defer { server.stop() }

        let url = URL(string: "http://127.0.0.1:\(server.port)/test")!
        var request = URLRequest(url: url)
        request.setValue("Bearer \(server.token)", forHTTPHeaderField: "Proxy-Authorization")
        request.setValue("Basic dXNlcjpwYXNz", forHTTPHeaderField: "Authorization")

        let (_, response) = try await URLSession.shared.data(for: request)
        let http = try #require(response as? HTTPURLResponse)

        #expect(http.statusCode == 200)
        #expect(http.value(forHTTPHeaderField: "X-Received-Auth") == "Basic dXNlcjpwYXNz")
    }

    // MARK: - CORS Preflight Auth Exemption

    @Test("preflight without token is answered under permissive CORS")
    func preflightWithoutTokenIsAnsweredUnderPermissiveCORS() async throws {
        // A preflight cannot carry credentials, so it is exempt from
        // authentication — and the library answers it itself, so the exemption
        // never reaches the handler.
        let server = try await HTTPServer.start(
            name: "auth-test",
            requiresAuthentication: true,
            cors: .permissive
        ) { _ in
            HTTPResponse(status: 200, body: Data("OK\n".utf8))
        }
        defer { server.stop() }

        let raw = "OPTIONS /test HTTP/1.1\r\nHost: 127.0.0.1\r\nAccess-Control-Request-Method: GET\r\n\r\n"
        let response = try await sendRaw(raw, toPort: server.port)
        #expect(response.hasPrefix("HTTP/1.1 204"))
    }

    @Test("preflight without token is authenticated without a CORS policy")
    func preflightWithoutTokenRequiresAuthWithoutCORS() async throws {
        // Without a policy to answer it, a preflight would reach the handler,
        // so the exemption would be an unauthenticated way in.
        let server = try await HTTPServer.start(
            name: "auth-test",
            requiresAuthentication: true
        ) { _ in
            HTTPResponse(status: 200, body: Data("OK\n".utf8))
        }
        defer { server.stop() }

        let raw = "OPTIONS /test HTTP/1.1\r\nHost: 127.0.0.1\r\nAccess-Control-Request-Method: GET\r\n\r\n"
        let response = try await sendRaw(raw, toPort: server.port)
        #expect(response.hasPrefix("HTTP/1.1 407"))
    }

    @Test("OPTIONS without Access-Control-Request-Method is not a preflight and returns 407")
    func nonPreflightOptionsWithoutTokenReturns407() async throws {
        // `canUser` issues a deliberate `OPTIONS` to read the `Allow` header. It
        // is a request the client made on its own behalf, so it carries the
        // token and must be authenticated like any other — the exemption covers
        // preflights, which cannot carry credentials, and nothing else.
        //
        // Under `.permissive` specifically: that is the only policy where the
        // exemption exists at all, so it is the only one where dropping the
        // `Access-Control-Request-Method` test would let this request through
        // unauthenticated. Under `.none` the assertion would hold no matter
        // what `isPreflight` returned.
        let server = try await HTTPServer.start(
            name: "auth-test",
            requiresAuthentication: true,
            cors: .permissive
        ) { _ in
            HTTPResponse(status: 200, body: Data("OK\n".utf8))
        }
        defer { server.stop() }

        let raw = "OPTIONS /test HTTP/1.1\r\nHost: 127.0.0.1\r\n\r\n"
        let response = try await sendRaw(raw, toPort: server.port)
        #expect(response.hasPrefix("HTTP/1.1 407"))
    }

    @Test("authenticated OPTIONS without Access-Control-Request-Method reaches the handler")
    func nonPreflightOptionsWithTokenReachesHandler() async throws {
        let server = try await HTTPServer.start(
            name: "auth-test",
            requiresAuthentication: true
        ) { request in
            HTTPResponse(status: 200, headers: [("Allow", request.parsed.method)], body: Data())
        }
        defer { server.stop() }

        let raw = "OPTIONS /test HTTP/1.1\r\nHost: 127.0.0.1\r\nRelay-Authorization: Bearer \(server.token)\r\n\r\n"
        let response = try await sendRaw(raw, toPort: server.port)
        #expect(response.hasPrefix("HTTP/1.1 200"))
        #expect(response.contains("Allow: OPTIONS"))
    }

    @Test("permissive CORS answers a preflight itself but forwards a deliberate OPTIONS")
    func permissiveCORSDistinguishesPreflightFromOptions() async throws {
        let server = try await HTTPServer.start(
            name: "auth-test",
            requiresAuthentication: true,
            cors: .permissive
        ) { _ in
            HTTPResponse(status: 200, headers: [("Allow", "GET, POST")], body: Data())
        }
        defer { server.stop() }

        let preflight = "OPTIONS /test HTTP/1.1\r\nHost: 127.0.0.1\r\nAccess-Control-Request-Method: POST\r\n\r\n"
        #expect(try await sendRaw(preflight, toPort: server.port).hasPrefix("HTTP/1.1 204"))

        // Without the preflight header the request belongs to the handler; the
        // library answering it with its own 204 would swallow the `Allow`
        // header `canUser` exists to read.
        let deliberate = "OPTIONS /test HTTP/1.1\r\nHost: 127.0.0.1\r\nRelay-Authorization: Bearer \(server.token)\r\n\r\n"
        let response = try await sendRaw(deliberate, toPort: server.port)
        #expect(response.hasPrefix("HTTP/1.1 200"))
        #expect(response.contains("Allow: GET, POST"))
    }

    @Test("GET without token still returns 407 (only a preflight is exempt)")
    func getWithoutTokenStillReturns407() async throws {
        let server = try await HTTPServer.start(
            name: "auth-test",
            requiresAuthentication: true
        ) { _ in
            HTTPResponse(status: 200, body: Data("OK\n".utf8))
        }
        defer { server.stop() }

        let url = URL(string: "http://127.0.0.1:\(server.port)/test")!
        let (_, response) = try await URLSession.shared.data(for: URLRequest(url: url))
        let http = try #require(response as? HTTPURLResponse)

        #expect(http.statusCode == 407)
    }

    // MARK: - Delegate Error Bodies

    @Test("a delegate's body answers a refusal the handler never sees")
    func delegateBodyAnswersRefusal() async throws {
        // A client that parses every response the same way — api-fetch reads
        // them all as JSON — reports a text/plain refusal as a parse failure,
        // losing the reason it was refused.
        let server = try await HTTPServer.start(
            name: "error-body-test",
            requiresAuthentication: true,
            delegate: JSONErrorDelegate()
        ) { _ in
            HTTPResponse(status: 200, body: Data("OK\n".utf8))
        }
        defer { server.stop() }

        let raw = "GET /test HTTP/1.1\r\nHost: 127.0.0.1\r\n\r\n"
        let response = try await sendRaw(raw, toPort: server.port)

        #expect(response.hasPrefix("HTTP/1.1 407"))
        #expect(response.contains("Content-Type: application/json"))
        #expect(response.contains(#"{"code":"refused"}"#))
        // The status and the protocol's own headers stay the server's: a
        // delegate supplies the payload, never the semantics.
        #expect(response.contains("Proxy-Authenticate: Bearer"))
    }

    @Test("a refusal falls back to the reason phrase without a delegate")
    func refusalWithoutDelegateIsPlainText() async throws {
        let server = try await HTTPServer.start(
            name: "error-body-test",
            requiresAuthentication: true
        ) { _ in
            HTTPResponse(status: 200, body: Data("OK\n".utf8))
        }
        defer { server.stop() }

        let raw = "GET /test HTTP/1.1\r\nHost: 127.0.0.1\r\n\r\n"
        let response = try await sendRaw(raw, toPort: server.port)

        #expect(response.hasPrefix("HTTP/1.1 407"))
        #expect(response.contains("Content-Type: text/plain"))
        #expect(response.contains("Proxy Authentication Required"))
    }

    // MARK: - Content-Length Requirement

    @Test("POST without Content-Length returns 411")
    func postWithoutContentLengthReturns411() async throws {
        let server = try await HTTPServer.start(
            name: "length-test",
            requiresAuthentication: true
        ) { _ in
            HTTPResponse(status: 200, body: Data("OK\n".utf8))
        }
        defer { server.stop() }

        // URLSession always adds Content-Length, so use a raw TCP connection.
        let raw = "POST /test HTTP/1.1\r\nHost: 127.0.0.1\r\nProxy-Authorization: Bearer \(server.token)\r\n\r\n"
        let response = try await sendRaw(raw, toPort: server.port)
        #expect(response.hasPrefix("HTTP/1.1 411"))
    }

    @Test("GET without Content-Length returns 200 (body not expected)")
    func getWithoutContentLengthReturns200() async throws {
        let server = try await HTTPServer.start(
            name: "length-test",
            requiresAuthentication: true
        ) { _ in
            HTTPResponse(status: 200, body: Data("OK\n".utf8))
        }
        defer { server.stop() }

        let url = URL(string: "http://127.0.0.1:\(server.port)/test")!
        var request = URLRequest(url: url)
        request.setValue("Bearer \(server.token)", forHTTPHeaderField: "Proxy-Authorization")

        let (_, response) = try await URLSession.shared.data(for: request)
        let http = try #require(response as? HTTPURLResponse)
        #expect(http.statusCode == 200)
    }

    @Test("POST with Content-Length returns 200")
    func postWithContentLengthReturns200() async throws {
        let server = try await HTTPServer.start(
            name: "length-test",
            requiresAuthentication: true
        ) { _ in
            HTTPResponse(status: 200, body: Data("OK\n".utf8))
        }
        defer { server.stop() }

        let url = URL(string: "http://127.0.0.1:\(server.port)/test")!
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.httpBody = Data("hello".utf8)
        request.setValue("Bearer \(server.token)", forHTTPHeaderField: "Proxy-Authorization")

        let (_, response) = try await URLSession.shared.data(for: request)
        let http = try #require(response as? HTTPURLResponse)
        #expect(http.statusCode == 200)
    }

    // MARK: - Auth Disabled

    @Test("authentication is disabled when requiresAuthentication is false")
    func authDisabledPassesThrough() async throws {
        let server = try await HTTPServer.start(
            name: "auth-test",
            requiresAuthentication: false
        ) { _ in
            HTTPResponse(status: 200, body: Data("OK\n".utf8))
        }
        defer { server.stop() }

        let url = URL(string: "http://127.0.0.1:\(server.port)/test")!
        let (_, response) = try await URLSession.shared.data(for: URLRequest(url: url))
        let http = try #require(response as? HTTPURLResponse)

        #expect(http.statusCode == 200)
    }
    // MARK: - Helpers

    /// Sends a raw HTTP request over TCP and returns the response string.
    ///
    /// Needed because URLSession always adds Content-Length automatically.
    private func sendRaw(_ request: String, toPort port: UInt16) async throws -> String {
        let connection = NWConnection(
            host: .ipv4(.loopback),
            port: NWEndpoint.Port(rawValue: port)!,
            using: .tcp
        )
        defer { connection.cancel() }

        try await withCheckedThrowingContinuation { (cont: CheckedContinuation<Void, any Swift.Error>) in
            connection.stateUpdateHandler = { state in
                switch state {
                case .ready:
                    connection.stateUpdateHandler = nil
                    cont.resume()
                case .failed(let error):
                    connection.stateUpdateHandler = nil
                    cont.resume(throwing: error)
                default:
                    break
                }
            }
            connection.start(queue: .global())
        }

        try await withCheckedThrowingContinuation { (cont: CheckedContinuation<Void, any Swift.Error>) in
            connection.send(content: Data(request.utf8), completion: .contentProcessed { error in
                if let error { cont.resume(throwing: error) } else { cont.resume() }
            })
        }

        return try await withCheckedThrowingContinuation { cont in
            connection.receive(minimumIncompleteLength: 1, maximumLength: 8192) { data, _, _, error in
                if let error {
                    cont.resume(throwing: error)
                } else {
                    cont.resume(returning: String(data: data ?? Data(), encoding: .utf8) ?? "")
                }
            }
        }
    }
}

/// Answers every server-generated error with a JSON body, the way a consumer
/// whose client parses all responses as JSON would.
private final class JSONErrorDelegate: HTTPServerDelegate {
    func errorBody(for error: HTTPServerError) -> HTTPErrorBody? {
        HTTPErrorBody(contentType: "application/json", data: Data(#"{"code":"refused"}"#.utf8))
    }
}

#endif // canImport(Network)
